sql: process column add/drop and index add in many chunks

vivekmenezes opened the pull request:
This adds in chunking to column truncation, column backfilling,
and index backfilling. Index truncation is still executed
in one transaction because it is supported nicely through
a DelRange() KV API call. We might need to change that later.

Also noteworthy there is no checkpointing in this PR. Thats coming soon.
#3274

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6056)

<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 212 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdvNvLBLSUw-Y3BX2q:-KFdvNvLBLSUw-Y3BX2r:-637044261) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L212)):*
nit: this might be more readable without so much logic inside the closure.

```go
var tableDesc *TableDescriptor
if pErr := sc.db.Txn(func(txn *client.Txn) *roachpb.Error {
  var pErr roachpb.Error
  tableDesc, pErr = getTableDescAtVersion(txn, sc.tableID, version)
  return pErr
}); pErr != nil {
  return span{}, pErr
}
prefix := roachpb.Key(MakeIndexKeyPrefix(tableDesc.ID, tableDesc.PrimaryIndex.ID))
return span{
  start: prefix,
  end:   prefix.PrefixEnd(),
}, nil
```

---

*[sql/backfill.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdwc5PTCzq1Ht_V5Oc:-KFdwc5PTCzq1Ht_V5Od:-1560678760) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L225)):*
it might also make sense to make this value dynamically adjust

---

*[sql/backfill.go, line 312 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdy3XvAbgoHjg7u2Rz:-KFdy3XvAbgoHjg7u2S-:-526019075) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L312)):*
seems like this is going to allocate quite a bit - you only need this for the very last key, right?

Not sure if it matters; perhaps it's time to add some benchmarks here.

---

*[sql/backfill.go, line 369 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxDWmJA0elcSJxaS5:-KFdxDWmJA0elcSJxaS6:-1479902849) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L369)):*
seems like you can return pErr unconditionally here

---

*[sql/backfill.go, line 420 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxHX7zs1lq3wxZyRX:-KFdxHX7zs1lq3wxZyRY:-2106563504) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L420)):*
ditto re doing this dynamically

---

*[sql/backfill.go, line 509 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdyJ03ZAt3NFp9g0FK:-KFdyJ03ZAt3NFp9g0FL:704750502) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L509)):*
do you need this check? doesn't the batch already no-op if it's empty?

---

*[sql/backfill.go, line 514 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxOK6NGhWwtx7T2-e:-KFdxOK6NGhWwtx7T2-f:-380945736) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L514)):*
Is this a question, or a sentence? Questions typically end in question marks.

---

*[sql/backfill.go, line 524 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxT9lL1ee9zYLZD0A:-KFdxT9lL1ee9zYLZD0B:1933868149) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L524)):*
ditto re unconditional return

---

*[sql/schema_changer_test.go, line 452 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdutcLdl1KZQ6dqEIW:-KFdutcLdl1KZQ6dqEIX:-351429821) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L452)):*
padding the stats?

---

*[sql/schema_changer_test.go, line 525 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdu2edKay1nJp316jG:-KFdu2edKay1nJp316jH:121979970) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L525)):*
yoda condition; make it `x != 1.4`. Also, where did this magic number come from?

---

*[sql/schema_changer_test.go, line 539 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFduFkgOiHVdcY6N20a:-KFduFkgOiHVdcY6N20b:269130074) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L539)):*
why 4? can you make this more self-documenting by avoiding magic numbers?

---

*[sql/schema_changer_test.go, line 542 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFduRl3hG5cNjk9788I:-KFduMsa4N6IIunn_OG3:571202848) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L542)):*
why don't you use `i` (after renaming it to rowNum, or something similar) instead of max_value + 1?

---

*[sql/schema_changer_test.go, line 548 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFduLUwbMXDk6Jr-zrM:-KFduLUwbMXDk6Jr-zrN:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L548)):*
ditto

---

*[sql/schema_changer_test.go, line 551 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFduXo3nraNFeJtaPbX:-KFduXo4Oja5jnZ0Yzmh:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/schema_changer_test.go#L551)):*
ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Turns out this PR is actually broken. Let me fix it incorporate all your feedback and repush.

vivekmenezes committed:
sql: chunking of schema change backfill operations

This adds in chunking to column truncate, column backfill,
and index backfill. Index truncate is still implemented
as one transaction because it is supported nicely through
a DelRange() KV API call. We might need to change that later.

There is no checkpointing in this commit; that's coming later.

vivekmenezes commented:
Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 212 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdvNvLBLSUw-Y3BX2q:-KGDVy76T_Cxaeb9ccAX:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L212)):*
done

---

*[sql/backfill.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdwc5PTCzq1Ht_V5Oc:-KGDW08bTbU83rx-69df:732521462) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L225)):*
I'd like to have these schema changes run really slow so that they are hardly noticeable, but I suppose someone will eventually ask for that speed up this schema change feature.

---

*[sql/backfill.go, line 312 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdy3XvAbgoHjg7u2Rz:-KGDX6joVdsQYfRzZpi2:-1479716845) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L312)):*
Moved the call to PrefixEnd() to the end so that it's called once. I think it will be valuable eventually to add benchmarks for the OLTP operations in the midst of these schema changes.

---

*[sql/backfill.go, line 369 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxDWmJA0elcSJxaS5:-KGDXjHPSsiUKkrWZ2Wu:-1743017777) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L369)):*
not anymore

---

*[sql/backfill.go, line 509 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdyJ03ZAt3NFp9g0FK:-KGDYetlNgf7fYs6M7kI:803535666) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L509)):*
good point

---

*[sql/backfill.go, line 514 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxOK6NGhWwtx7T2-e:-KGDYpAB_NT_QbbdyFDp:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L514)):*
done

---

*[sql/backfill.go, line 524 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxT9lL1ee9zYLZD0A:-KGDXanzUEc48EHn-Kcv:-1743017777) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L524)):*
not anymore

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056)*
<!-- Sent from Reviewable.io -->

tamird commented:
Did you mean to remove all the tests you added?

---

Reviewed 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdwc5PTCzq1Ht_V5Oc:-KGDvARlWBnrApPuXTHh:1642922842) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L225)):*
To clarify, I'm only suggesting an amendment to this TODO.

---

*[sql/backfill.go, line 376 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGDvRpEqSrBJ_xp2knV:-KGDvRpFJxO_TukG3BYp:-1783053805) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4ac2c56b53e8950273ff965c74f0cadcde39c67/sql/backfill.go#L376)):*
can this go in the `for`? i think it'd be clearer there.

---

*[sql/backfill.go, line 532 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGDvicHBQIjmrNmXahL:-KGDvicHBQIjmrNmXahM:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4ac2c56b53e8950273ff965c74f0cadcde39c67/sql/backfill.go#L532)):*
ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-:-KGDvwfYXoP-6EpAJtNW:1425255032)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Yeah those tests were moved to their own PR. They went through major change and are now all checked in. This PR is only increasing maxValues in the tests.

---

Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdwc5PTCzq1Ht_V5Oc:-KGKZF1yn0gdNcTyiHVC:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L225)):*
done

---

*[sql/backfill.go, line 420 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KFdxHX7zs1lq3wxZyRX:-KGK_X77KbOonmEGNsQN:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/caaee68e14e7d6d78d69b2574accc4e99c05a7c1/sql/backfill.go#L420)):*
done

---

*[sql/backfill.go, line 376 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGDvRpEqSrBJ_xp2knV:-KGK_VzZo4XqwUiLgcnY:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4ac2c56b53e8950273ff965c74f0cadcde39c67/sql/backfill.go#L376)):*
done

---

*[sql/backfill.go, line 532 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGDvicHBQIjmrNmXahL:-KGK_YJ-hinneaDaoDLU:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4ac2c56b53e8950273ff965c74f0cadcde39c67/sql/backfill.go#L532)):*
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-:-KGKY_GW3qnVS2oP62ZQ:-134460361)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM with no sleep :)

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 223 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNQenS3rtpBfsNJtgF:-KGNQenTzi14OT7u9V9S:1981384309) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L223)):*
looks like you've added the sleep already. As discussed, I think we should just remove it.

---

*[sql/backfill.go, line 278 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNRtHVKhbNHUd-KK01:-KGNRtHVKhbNHUd-KK02:-864397187) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L278)):*
I think it'd be good to move this inner Txn to its own function. Kinda hard to read otherwise - so much indentation.

---

*[sql/backfill.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNTaztEQZAIq_ae7Sh:-KGNTaztEQZAIq_ae7Si:1220966691) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L287)):*
I'd add a comment that using a cursor like this with multiple txns is cool because new cols and indexes are in a write-only state, so we're not gonna miss any new rows.

---

*[sql/backfill.go, line 344 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNS7gb5MF9XIRbPrH0:-KGNS7gb5MF9XIRbPrH1:-1108557839) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L344)):*
nit: increase this log level. 2 is what we use for all debugging, this will be super verbose.

---

*[sql/backfill.go, line 422 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNe-1H4ZaXt3V26oNQ:-KGNe-1H4ZaXt3V26oNR:400751055) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L422)):*
the sleep is already added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-:-KGNe406ZBG2rDFbGe9L:1258703300)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 344 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056#-KGNS7gb5MF9XIRbPrH0:-KGNhJnB18l1KnePGVSH:-140462197) ([raw file](https://github.com/cockroachdb/cockroach/blob/59daa6cc8d9063f5225f2fd8703b518abf8bbc0a/sql/backfill.go#L344)):*
I believe log.V(2) is used throughout the sql package for this kind of logging. If you want to change it across the board, I would suggest doing so in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6056)*
<!-- Sent from Reviewable.io -->
